src/black/__init__.py

```diff
@@ -1061,6 +1060,8 @@ def f(
             current_line, mode=mode, features=split_line_features
         ):
             dst_contents.append(str(line))
+    if not dst_contents:
+        return "\n"
```

Comment:
What do other formatters do here when running on windows vs. POSIX?

Should we (if found) leave Windows like terminators alone? (to leave windows test files alone? i.e. `"\r\n"`)

Reply:
That'd be good for consistency with nonempty files 😅

Reply:
>Should we (if found) leave Windows line terminators alone?

I believe this patch doesn't change Black's behavior with regard to existing line terminators. 

Replacing `"\n"` with `os.linesep` here and in corresponding tests should make this behave in a more native way on Windows, right?
```suggestion
    if not dst_contents:
        return os.linesep
```

Reply:
>I believe this patch doesn't change Black's behavior with regard to existing line terminators.

Ah, but of course it does. A whitespace-only file with Windows line terminator first becomes empty, and is then changed into a single Unix linefeed. Hm.

Reply:
Yeah - This could also induce changes all the time if you have developers on Windows vs. POSIX ... i.e. file gets reformatted on Windows, then committed, then a POSIX dev will reformat and commit (we could have a constant battle)

We will just have to pick the one - which I think `\n` is more accepted in Python source files. Editors have been taught to leave this alone by default usually I think.

Reply:
Surely git will take care of changing the endings though!

Reply:
Some discussion about the line endings in #1741

Reply:
I think I agree with @zsol's [comment](https://github.com/psf/black/issues/1741#issuecomment-703775430):
>IIRC we have considered this, but ultimately decided to be extra careful and preserve the original line endings, as we felt that's more in the domain of the version control software being used. The only opinion Black has on this matter is that line endings should be consistent (which is a fairly basic requirement :) )

So for zero-length files which Black turns into a single line terminator, I see three viable options:
1. always `\n`
2. `os.linesep`, so `\n` on Unix and `\r\n` on Windows
3. somehow sniff or remember the line terminator in neighboring files and use that

I never develop on Windows, so I don't know if it's a more common case to see CR or CRLF line terminators in Python projects when using most common tooling on that platform.

Reply:
I think it should be the system line ending, because git can be used for conversion. I don't really know if it is more common to still use `\n`, but at least I've not encountered it. Might have missed it though as well, if editors are transparent about that.

Reply:
> What do other formatters do here when running on windows vs. POSIX?
>
> Should we (if found) leave Windows line terminators alone? (i.e. "\r\n")

I do not have a windows machine at my disposal but I tested a few formatters from different languages:

* Yapf: Leaves a completely empty file as is,  a whitespace only file without newline gets converted to `\n`, a file with a single newline (`\n` or `\r\n`) is preserved the way it was before and a file with multiple newlines is truncated to a `\n` (even if the newlines were `\r\n`).
* Prettier (for JS): Defaults to `\n` for each platform.
* Rustfmt (for Rust): Seems to convert an empty file to `\n`, otherwise identical to yapf.

Personally I think it would be the best compromise is to leave a completely empty file as is and a file with non-newline whitespace should be converted to a completely empty file. For all other files we can infer the line ending and preserve it.

Reply:
So the "votes" so far for what zero-length files should become on Windows go like:
- **0.5** for keep empty: @JelleZijlstra (in issue 2382)
- **2** for LF: @JelleZijlstra, @septatrix (in issue 2382), @cooperlees (in this issue)
- **1.5** for CRLF: @cooperlees (?), @felix-hilden (in this issue)

I'm counting half a vote for those who are ok with two different alternatives. @JelleZijlstra & @cooperlees, do you want to choose just one preference based on the discussion above? Is "voting" even a good way to reach a design decision in Black, or do we have a BDFL to dictate the choice?

Reply:
Found one linting issue with empty files:
```bash
$ echo >/tmp/empty.py
$ flake8 /tmp/empty.py
/tmp/empty.py:1:1: W391 blank line at end of file
$ pylint /tmp/empty.py      
************* Module empty
/tmp/empty.py:1:0: C0305: Trailing newlines (trailing-newlines)
```

So [Flake8](https://flake8.pycqa.org/) and [Pylint](https://pylint.org/) seem to think empty source files should be zero-length instead of containing any line terminators! I'll add these as "voters" above.

Reply:
I agree with Flake8 and Pylint FWIW

Reply:
I now changed the implementation to emit a
- zero-length file for empty or all-whitespace input but now line terminators
- single line terminator for all-whitespace input with at least one line terminator, preserving LF/CRLF line terminators

Reply:
I couldn't figure out why there were linting errors in the GitHub Actions workflow for files this branch didn't even touch.